src/codegen.c

```diff
@@ -765,55 +805,79 @@ void codegen_lower(CodegenContext *context) {
     case CG_FMT_x86_64_GAS:
       codegen_lower_x86_64(context);
       break;
+    case CG_FMT_IR:
+      codegen_lower_ir_backend(context);
+      break;
     default:
       TODO("Handle %d code generation format.", context->format);
   }
 }
 
 void codegen_emit(CodegenContext *context) {
   switch (context->format) {
-  case CG_FMT_x86_64_GAS:
-    codegen_emit_x86_64(context);
-    break;
-  default:
-    TODO("Handle %d code generation format.", context->format);
+    case CG_FMT_x86_64_GAS:
+      codegen_emit_x86_64(context);
+      break;
+    case CG_FMT_IR:
```

Comment:
For some reason, when passing `gcd.ir --format x86_64_gas` to the compiler, `code.S` has IR output written to it.

Reply:
```
.section .text

.global main () global {
    %0 = imm 88
    %1 = imm 32
    tail call gcd (%0, %1)
    unreachable
}

defun gcd (%0, %1) global leaf pure {
    %2 = eq %0, %1
    br.cond %2, bb5, bb2
bb2:
    %3 = lt %0, %1
    br.cond %3, bb3, bb4
bb3:
    %4 = sub %1, %0
    tail call gcd (%0, %4)
    unreachable
bb4:
    %5 = sub %0, %1
    tail call gcd (%5, %1)
    unreachable
bb5:
    ret %0
}


.global gcd (%0, %1) global leaf pure {
    %2 = eq %0, %1
    br.cond %2, bb5, bb2
bb2:
    %3 = lt %0, %1
    br.cond %3, bb3, bb4
bb3:
    %4 = sub %1, %0
    tail call gcd (%0, %4)
    unreachable
bb4:
    %5 = sub %0, %1
    tail call gcd (%5, %1)
    unreachable
bb5:
    ret %0
}



main () global {
    %0 = imm 88
    %1 = imm 32
    tail call gcd (%0, %1)
    unreachable
}

defun gcd (%0, %1) global leaf pure {
    %2 = eq %0, %1
    br.cond %2, bb5, bb2
bb2:
    %3 = lt %0, %1
    br.cond %3, bb3, bb4
bb3:
    %4 = sub %1, %0
    tail call gcd (%0, %4)
    unreachable
bb4:
    %5 = sub %0, %1
    tail call gcd (%5, %1)
    unreachable
bb5:
    ret %0
}

:
    push %rbp
    mov %rsp, %rbp
    sub $40, %rsp
.L0:
    mov $88, %rcx
    mov $32, %rdx
    mov %rbp, %rsp
    pop %rbp
    jmp gcd (%0, %1) global leaf pure {
    %2 = eq %0, %1
    br.cond %2, bb5, bb2
bb2:
    %3 = lt %0, %1
    br.cond %3, bb3, bb4
bb3:
    %4 = sub %1, %0
    tail call gcd (%0, %4)
    unreachable
bb4:
    %5 = sub %0, %1
    tail call gcd (%5, %1)
    unreachable
bb5:
    ret %0
}



gcd (%0, %1) global leaf pure {
    %2 = eq %0, %1
    br.cond %2, bb5, bb2
bb2:
    %3 = lt %0, %1
    br.cond %3, bb3, bb4
bb3:
    %4 = sub %1, %0
    tail call gcd (%0, %4)
    unreachable
bb4:
    %5 = sub %0, %1
    tail call gcd (%5, %1)
    unreachable
bb5:
    ret %0
}

:
    push %rbp
    mov %rsp, %rbp
    sub $40, %rsp
.L1:
    cmp %rdx, %rcx
    mov $0, %rax
    sete %al
    test %rax, %rax
    jz .L2
    jmp .L5
.L2:
    cmp %rdx, %rcx
    mov $0, %rax
    setl %al
    test %rax, %rax
    jz .L4
    jmp .L3
.L3:
    sub %rcx, %rdx
    mov %rbp, %rsp
    pop %rbp
    jmp gcd (%0, %1) global leaf pure {
    %2 = eq %0, %1
    br.cond %2, bb5, bb2
bb2:
    %3 = lt %0, %1
    br.cond %3, bb3, bb4
bb3:
    %4 = sub %1, %0
    tail call gcd (%0, %4)
    unreachable
bb4:
    %5 = sub %0, %1
    tail call gcd (%5, %1)
    unreachable
bb5:
    ret %0
}


.L4:
    sub %rdx, %rcx
    mov %rbp, %rsp
    pop %rbp
    jmp gcd (%0, %1) global leaf pure {
    %2 = eq %0, %1
    br.cond %2, bb5, bb2
bb2:
    %3 = lt %0, %1
    br.cond %3, bb3, bb4
bb3:
    %4 = sub %1, %0
    tail call gcd (%0, %4)
    unreachable
bb4:
    %5 = sub %0, %1
    tail call gcd (%5, %1)
    unreachable
bb5:
    ret %0
}


.L5:
    mov %rcx, %rax
    mov %rbp, %rsp
    pop %rbp
    ret
```

This *can't* be right

Reply:
Er, that’s... probably not intended? I’ll take a look at that haha.

Reply:
Providing `strndup` on Windows *seems* to have fixed this too